File: androidmvp/src/main/java/com/hsc/vince/androidmvp/base/MvpActivity.java
package com.hsc.vince.androidmvp.base;

import android.os.Bundle;
import android.support.annotation.Nullable;

/**
 * <p>作者：黄思程  2018/4/12 10:06
 * <p>邮箱：codinghuang@163.com
 * <p>作用：
 * <p>描述：带网络请求的Activity
 */
public abstract class MvpActivity<P extends BasePresenter> extends BaseActivity {
    protected P mPresenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        mPresenter = createPresenter();
        if (mPresenter == null) {
            throw new NullPointerException("presenter cannot be nill");
        }

        super.onCreate(savedInstanceState);
    }

    protected abstract P createPresenter();

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (mPresenter != null) {
            mPresenter.detachView();
        }
    }

    @Override
    public void onBackPressed() {
        finish();

        if (mPresenter != null) {
            mPresenter.onUnsubscribe();
        }
    }
}

File: news/src/main/java/com/vince/news/base/MvpActivity.kt
package com.vince.news.base

import android.os.Bundle
import android.view.Window

/**
 * <p>作者：黄思程  2017/11/16 16:48
 * <p>邮箱：huangsicheng@gouuse.cn
 * <p>作用：
 * <p>描述：带网络请求的Activity
 */
abstract class MvpActivity<P : BasePresenter<*>> : BaseActivity() {
    protected lateinit var mvpPresenter: P

    override fun onCreate(savedInstanceState: Bundle?) {
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        mvpPresenter = createPresenter()

        super.onCreate(savedInstanceState)
    }

    override fun onDestroy() {
        super.onDestroy()

        mvpPresenter.detachView()
    }

    override fun onBackPressed() {
        super.onBackPressed()

        mvpPresenter.unsubscribe()
    }

    /**
     * @return 创建Presenter
     * */
    protected abstract fun createPresenter(): P

}